R/compat-downstream-dep.R

```diff
@@ -82,20 +102,39 @@ check_downstream <- function(ver, ..., with_rlang = requireNamespace("rlang")) {
   pkg_ver <- utils::packageVersion(pkg)
 
   header <- sprintf("As of %s %s, %s must be at least version %s.", pkg, pkg_ver, dep_pkg, dep_min)
-  body <- c(
-    x = sprintf("%s %s is too old for %s %s.", dep_pkg, curr_ver, pkg, pkg_ver),
-    .rlang_downstream_howto_reinstall_msg(dep_pkg)
-  )
+  info_pkg <- c(x = sprintf("%s %s is too old for %s %s.", dep_pkg, curr_ver, pkg, pkg_ver))
+  info_install <- .rlang_downstream_howto_reinstall_msg(dep_pkg)
 
   if (with_rlang) {
     # Don't use `::` because this is also called from rlang's onLoad
     # hook and exports are not initialised at this point
     warn <- get("warn", envir = asNamespace("rlang"))
-    warn(c(header, body))
+    inform <- get("inform", envir = asNamespace("rlang"))
+    is_interactive <- get("is_interactive", envir = asNamespace("rlang"))
+  } else {
+    format_msg <- function(x) paste(x, collapse = "\n")
+    warn <- function(msg) warning(format_msg(msg), call. = FALSE)
+    inform <- function(msg) message(format_msg(msg))
+    is_interactive <- function() interactive() && !isFALSE(getOption("rlang_interactive"))
+  }
+
+  if (!is_interactive()) {
+    warn(c(header, info_pkg, info_install))
+    return()
+  }
+
+  prompt <- c("!" = sprintf("Would you like to update the package %s now?", dep_pkg))
+  inform(c(header, info_pkg, prompt))
+
+  if (utils::menu(c("Yes", "No")) != 1) {
+    inform("Set `options(rlib_downstream_check = FALSE)` to disable this prompt.")
+    return()
+  }
+
+  if (is_installed("pak")) {
```

Comment:
Is this the same code path as `check_installed()` or different?

Reply:
This is the same code but not the same code path.